### DISCUSSION TITLE
Can the TestCluster be used to obtain items from the service collection?

### DISCUSSION THREAD

**Discussion by macfma01** [dotnet/orleans#8732] [Q&A] [answered]
I've been successfully using the test cluster to get grain instances for testing. The cluster was configured with several services in it's service collection. The service interfaces are injected into the grain classes. There is a `ServiceProvider` property on the test cluster instance and it has a `GetService` method so I thought that I could obtain these services this way. When I try I get null for whatever service I ask for. No problem getting grain instances so I know at least that part of it working.

Is it even possible to retrieve the services this way?

**Comment by runenilsenoe** [comment, marked as answer]
Indeed by passing config inheriting ISiloConfigurator.
There you can either mock or use the service directly

```
public sealed class ClusterFixture : IDisposable
{
    public TestCluster Cluster { get; } = new TestClusterBuilder()
        .AddSiloBuilderConfigurator<TestSiloConfigurations>()
        .Build();
    public ClusterFixture() => Cluster.Deploy();
    void IDisposable.Dispose() => Cluster.StopAllSilos();
}
public class TestSiloConfigurations : ISiloConfigurator
{
    public void Configure(ISiloBuilder siloBuilder)
    {
        siloBuilder.ConfigureServices(x =>
        {
            x.AddSingleton<IService,Service>();
            x.AddSingleton(new Mock<IService>().Object);
        });
    }
}`
